File: portal/src/lib/api/compliance.ts
import { apiClient } from "./client";
import type { Page, PageParams } from "./types";

export type CheckStatus = "pass" | "fail" | "unknown";
export type DeviceComplianceStatus = "compliant" | "at_risk" | "non_compliant" | "unknown";

export interface CheckResult {
  key: string;
  label: string;
  status: CheckStatus;
  detail: string;
  fix_action_id: string | null;
}

export interface DeviceCompliance {
  device_id: string;
  hostname: string;
  status: DeviceComplianceStatus;
  score: number;
  passed: number;
  failed: number;
  checks: CheckResult[];
}

export interface CheckBreakdown {
  key: string;
  label: string;
  passed: number;
  failed: number;
  unknown: number;
}

export interface UsbPosture {
  blocked: number;
  allowed: number;
  unknown: number;
}

export interface ComplianceSummary {
  total_devices: number;
  compliant: number;
  at_risk: number;
  non_compliant: number;
  unknown: number;
  score: number;
  checks: CheckBreakdown[];
  usb: UsbPosture;
}

export interface BannedSoftware {
  id: string;
  name: string;
  pattern: string;
  created_at: string;
}

export const getComplianceSummary = () =>
  apiClient.get<ComplianceSummary>("/compliance/summary").then((r) => r.data);

export const getComplianceDevices = (
  params: PageParams & { needs_attention?: boolean } = {},
) =>
  apiClient.get<Page<DeviceCompliance>>("/compliance/devices", { params }).then((r) => r.data);

export const getDeviceCompliance = (deviceId: string) =>
  apiClient.get<DeviceCompliance>(`/compliance/devices/${deviceId}`).then((r) => r.data);

export const listBannedSoftware = () =>
  apiClient.get<BannedSoftware[]>("/compliance/banned-software").then((r) => r.data);

export const addBannedSoftware = (name: string) =>
  apiClient.post<BannedSoftware>("/compliance/banned-software", { name }).then((r) => r.data);

export const removeBannedSoftware = (id: string) =>
  apiClient.delete(`/compliance/banned-software/${id}`).then((r) => r.data);

File: backend/app/schemas/compliance.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

CheckStatus = Literal["pass", "fail", "unknown"]
DeviceComplianceStatus = Literal["compliant", "at_risk", "non_compliant", "unknown"]


class BannedSoftwareCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)


class BannedSoftwareRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    pattern: str
    created_at: datetime


class CheckResult(BaseModel):
    key: str
    label: str
    status: CheckStatus
    detail: str
    # The remediation an admin can push to fix this check, if one applies.
    fix_action_id: str | None = None


class DeviceCompliance(BaseModel):
    device_id: uuid.UUID
    hostname: str
    status: DeviceComplianceStatus
    score: int  # 0–100, over the device's *known* checks
    passed: int
    failed: int
    checks: list[CheckResult]
    # The device's reported USB storage state; None until a reporting agent has beaten.
    usb_storage_blocked: bool | None = None


class CheckBreakdown(BaseModel):
    key: str
    label: str
    passed: int
    failed: int
    unknown: int


class UsbPosture(BaseModel):
    """How USB mass storage sits across the fleet, from what each agent last reported.
    `unknown` is devices no agent new enough to report it has checked in for yet."""
    blocked: int
    allowed: int
    unknown: int


class ComplianceSummary(BaseModel):
    total_devices: int
    compliant: int
    at_risk: int
    non_compliant: int
    unknown: int
    score: int  # fleet compliance % = compliant / total
    checks: list[CheckBreakdown]
    usb: UsbPosture
